File: Concurrency/Java/print-in-order.java
class Foo {
    private Semaphore secondLock;
    private Semaphore thirdLock;
    public Foo() {
        secondLock = new Semaphore(0);
        thirdLock = new Semaphore(0);
    }

    public void first(Runnable printFirst) throws InterruptedException {
        printFirst.run();
        secondLock.release();
    }

    public void second(Runnable printSecond) throws InterruptedException {
        secondLock.acquire();
        printSecond.run();
        thirdLock.release();
    }

    public void third(Runnable printThird) throws InterruptedException {
        thirdLock.acquire();
        printThird.run();
    }
}

File: Concurrency/c++/print-in-order.cpp
#include <pthread.h> 
#include <semaphore.h> 

class Foo {
private:
    sem_t secondLock;
    sem_t thirdLock;
public:
    Foo() {
        sem_init(&secondLock, 0, 0); 
        sem_init(&thirdLock, 0, 0); 
    }
    
    ~Foo() {
        sem_destroy(&secondLock);
        sem_destroy(&thirdLock);
    }

    void first(function<void()> printFirst) {
        // printFirst() outputs "first". Do not change or remove this line.
        printFirst();
        sem_post(&secondLock);
    }

    void second(function<void()> printSecond) {
        sem_wait(&secondLock);
        // printSecond() outputs "second". Do not change or remove this line.
        printSecond();
        sem_post(&thirdLock);
    }

    void third(function<void()> printThird) {
        sem_wait(&thirdLock);
        // printThird() outputs "third". Do not change or remove this line.
        printThird();
    }
};
